Fix interval classes for MaskedArray and MaskedNDArray

larrybradley opened the pull request:
<!-- These comments are hidden when you submit the pull request,
so you do not need to remove them! -->

<!-- Please be sure to check out our contributing guidelines,
https://github.com/astropy/astropy/blob/main/CONTRIBUTING.md .
Please be sure to check out our code of conduct,
https://github.com/astropy/astropy/blob/main/CODE_OF_CONDUCT.md . -->

<!-- If you are new or need to be re-acquainted with Astropy
contributing workflow, please see
https://docs.astropy.org/en/latest/development/quickstart.html .
There is even a practical example at
https://docs.astropy.org/en/latest/development/git_edit_workflow_examples.html . -->

<!-- Please just have a quick search on GitHub to see if a similar
pull request has already been posted.
We have old closed pull requests that might provide useful code or ideas
that directly tie in with your pull request. -->

<!-- We have several automatic features that run when a pull request is open.
They can appear daunting but do not worry because maintainers will help
you navigate them, if necessary. -->

### Description
<!-- Provide a general description of what your pull request does.
Complete the following sentence and add relevant details as you see fit. -->

<!-- In addition please ensure that the pull request title is descriptive
and allows maintainers to infer the applicable subpackage(s). -->

<!-- READ THIS FOR MANUAL BACKPORT FROM A MAINTAINER:
Apply "skip-basebranch-check" label **before** you open the PR! -->

The interval classes currently drop the mask of `MaskedArray` input, thus returning incorrect `vmin` and `vmax` values.

Simple MWE:

```python
import numpy as np
from astropy.visualization import simple_norm

rng = np.random.default_rng(seed=0)
data = rng.random((50, 50))
data[0, :] = 1.e6 
data = np.ma.masked_where(data > 1, data)
norm = simple_norm(data, 'linear', percent=100)
plt.imshow(data, norm=norm)
plt.colorbar()
```

Current main:

![main](https://github.com/user-attachments/assets/d9bb09af-ac40-4939-8cc7-6ac6ae2407ff)


With this PR:

![pr](https://github.com/user-attachments/assets/05676aef-5174-4b6e-a58d-ce0a3826ecf4)


<!-- If the pull request closes any open issues you can add this.
If you replace <Issue Number> with a number, GitHub will automatically link it.
If this pull request is unrelated to any issues, please remove
the following line. -->


<!-- Optional opt-out -->

- [ ] By checking this box, the PR author has requested that maintainers do **NOT** use the "Squash and Merge" button. Maintainers should respect this when possible; however, the final decision is at the discretion of the maintainer that merges the PR.

mhvk commented on astropy/visualization/interval.py:
```diff
@@ -110,12 +138,7 @@ def get_limits(self, values):
             return self.vmin, self.vmax
 
-        # Make sure values is a Numpy array
-        values = np.asarray(values).ravel()
```
I think a simpler `np.asanyarray(...)` here should work, without any further extra code (bit of an aside, but: I don't see why you'd need `.ravel()`).

larrybradley replied:
`asanyarray` fails because later functions that operate on the data (like np.percentile) ignore the mask and give warnings:
`UserWarning: Warning: 'partition' will ignore the 'mask' of the MaskedArray.   We really need to extract the non-masked values.

I agree about the `ravel` -- I think we can drop that.

larrybradley replied:
@mhvk Should I add support for Astropy `Masked` too?  i.e., `values = values.unmasked[~values.mask]`

mhvk replied:
Ah, I even misread - I thought we were talking about `Masked`, which I think might be able to deal with `percentile`. You could use `astropy.utils.masked.get_data_and_mask` to get the relevant data and mask independent of masked class. (There may be some use in doing it differently by making everything `Masked` and using that for the finite stuff too, but a bit too late here to think that through...)

larrybradley commented:
Updated:  I also added support for Quantity and masked_Quantity inputs by simply dropping the units (mpl fails for Quantity inputs and Quantity vmin/vmax inputs).

mhvk commented:
@larrybradley - sorry to have been quiet on this (end of term here...). I still think the mask stuff is good, but I don't like just ignoring units. I realize we are in a bind here, since matplotlib doesn't support units for images, but I think it is then better to just let things fail, or to rewrite at least our scaler classes such that they treat units consistently. I'm not sure what is best, so my suggestion would be to just stick to the mask stuff here.

Since I also had a suggestion to ensure `MaskedNDArray` works, and asking someone to revert is annoying, I made a pull request to your branch that does it. Including that, I think all is OK (though we should squash & merge unless you want to reorganize the commits). See https://github.com/larrybradley/astropy/pull/1

mhvk commented:
p.s. If you agree with my idea of just doing masked arrays for now, we can open a new issue to remind us of the `Quantity` stuff.

larrybradley commented:
@mhvk I merged your PR.  I'm fine with excluding the masked Quantity stuff.

I did discover an annoying issue with unmasked Quantity that I'll open a separate issue/PR for.

mhvk reviewed (approved):
Looks good!

mhvk commented:
pre-commit.ci autofix

mhvk commented:
Hmm, autofix could not fix
```
astropy/modeling/tests/test_fitters.py:1481:9: B007 Loop control variable `i` not used within loop body
     |
1479 |         ctx2 = nullcontext()
1480 |
1481 |     for i in range(2 + ndim):
     |         ^ B007
1482 |         params = np.roll(params, 1, axis=0)
1483 |         with ctx1, ctx2:
     |
     = help: Rename unused `i` to `_i`
```
But what has that got to do with this PR?

larrybradley commented:
I have a fix here: https://github.com/astropy/astropy/pull/18030

This PR will need to be rebased.

larrybradley commented:
rebased

pllim commented:
> I did discover an annoying issue with unmasked Quantity that I'll open a separate issue/PR for.

Please link that issue/PR here, @larrybradley . Thanks!
